### PR TITLE
Mobility GHC927 Upgrade Final

### PR DESCRIPTION
**Pull request by arjunkathuria**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [x] Refactoring
- [x] Dependency updates

## Description
This PR enables the shared-kernel repository to build with GHC 9.2.7 and all its upgraded package-set and dependencies that come with the upgraded GHC927 package set, updated and rebased to work with the newest commit.

This is part of the larger efforts to upgrade and bring the entire nammayatri ecosystem on GHC 9.2.

To give a brief summary of changes, including but not limited to :- 
- Modifies the codebase to accomodate the updated juspay and internal dependencies, which were in-turn patched to be able to build with GHC 9.2.7 
- Handles and fixes all breaking changes across major version of repositories like Aeson v2+, Text 2+ etc.
- Updates the nix setup accordingly.
- It also upgrades the project tooling like Haskell-Language-Server (HLS) etc.

### Additional Changes
modifies the `flake.nix` and `flake.lock` for dependency upgrades

## Motivation and Context
GHC 9.2 upgrade. A full document of reasons and benefits is being prepared and will be shared internally across teams.

## How did you test it?

- The project builds on my local
- Passes all pre-commit checks
- The entire test-suite builds and passes successfully.

local rideflow testing was also done.

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by srid** [comment]
@vsaimanohar Could you try out this PR locally and make sure that everything works (cabal build, nix build, etc.)?

I'm asking because we've reverted the fix <https://github.com/nammayatri/nammayatri/pull/2930> for your old problem in this PR.

**Review by srid — APPROVED** [review]
Nix parts LGTM

File: flake.nix [hunk 1]
```diff
@@ -10,7 +12,11 @@
     prometheus-haskell.url = "github:juspay/prometheus-haskell/more-proc-metrics";
     prometheus-haskell.inputs.haskell-flake.follows = "common/haskell-flake";
 
-    euler-hs.url = "github:juspay/euler-hs/ag/open-source";
+    euler-hs = {
+      url = "github:juspay/euler-hs/ag/open-source";
+      inputs.nixpkgs.follows = "nixpkgs";
+      inputs.haskell-flake.follows = "haskell-flake";
```
Comment (srid) [inline flake.nix:18]: It is unfortunate that we can't use `"common/haskell-flake"` directly here. Is there an upstream issue for it?

**Comment by vsaimanohar** [comment]
> @vsaimanohar Could you try out this PR locally and make sure that everything works (cabal build, nix build, etc.)?
> 
> I'm asking because we've reverted the fix [nammayatri/nammayatri#2930](https://github.com/nammayatri/nammayatri/pull/2930) for your old problem in this PR.

@srid i have tested this PR 
before updating flake input to point to my forked prometheus PR:- 

nix build, cabal build are working fine

after updating flake input to point to my forked prometheus PR:-
nix build, cabal build are throwing error as sync from forked prometheus PR to shared-kernel is not getting applied

> src/Kernel/Tools/Metrics/Init.hs:55:7: error:
>     • Variable not in scope:
>         instrumentHandlerValueWithLabels

**Review by hkmangla — APPROVED** [review]

**Pull request merged**